pennylane/_device.py

```diff
@@ -285,41 +285,73 @@ def supports_operation(self, operation):
         Returns:
             bool: ``True`` iff supplied operation is supported
         """
-        if isinstance(operation, type) and issubclass(operation, Operation):
-            return operation.__name__ in self.operations
-        if isinstance(operation, str):
-            return operation in self.operations
+        op_name = operation
 
-        raise ValueError("The given operation must either be a pennylane.Operation class or a string.")
+        if issubclass(operation.__class__, Operation):
+            op_name = operation.name
+        elif isinstance(operation, type):
+            op_name = operation.__name__
+        elif not isinstance(op_name, str):
+            raise ValueError("The given operation must either be a pennylane.Operation class or a string.")
+
+        if op_name not in self.operations:
+            raise DeviceError("Gate {} not supported on device {}".format(op_name, self.short_name))
+
+        return True
 
     def supports_observable(self, observable):
         """Checks if an observable is supported by this device.
 
         Args:
-            operation (Observable,str): observable to be checked
+            observable (Observable,str): observable to be checked
 
         Returns:
             bool: ``True`` iff supplied observable is supported
         """
-        if isinstance(observable, type) and issubclass(observable, Observable):
-            return observable.__name__ in self.observables
-        if isinstance(observable, str):
-            return observable in self.observables
+        obs_name = observable
+
+        if issubclass(observable.__class__, Observable):
+            obs_name = observable.name
+        elif isinstance(observable, type):
+            obs_name = observable.__name__
+        elif not isinstance(obs_name, str):
+            raise ValueError("The given observable must either be a pennylane.Observable class or a string.")
+
+        if obs_name not in self.observables:
+            raise DeviceError("Observable {} not supported on device {}".format(obs_name, self.short_name))
 
-        raise ValueError("The given operation must either be a pennylane.Observable class or a string.")
+        return True
+
+    def supports_inverse(self):
+        """Checks if this device supports inverse operations.
+
+        Returns:
+            bool: ``True`` iff inverse operations are supported by the device
+        """
+        if "inverse_operations" not in self.capabilities() or not self.capabilities()["inverse_operations"]:
```

Comment:
I think (could be wrong) that this could be replaced with the `dict.get()` method:
```suggestion
        if self.capabilities().get("inverse_operations", False):
```

Reply:
That's true! Although then the entire statement needs to be negated (such that it raises an error):
``if not self.capabilities().get("inverse_operations", False):``

I'll change this and also apply the same change for the Tensor check in ``check_validity``